File: app/src/main/java/io/celox/xclip/MainActivity.java
package io.celox.xclip;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import io.celox.xclip.service.ClipboardService;
import io.celox.xclip.ui.ClipboardDialog;

public class MainActivity extends AppCompatActivity {

    private static final int PERMISSION_REQUEST_CODE = 100;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Check for notification permission (Android 13+)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS)
                    != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this,
                        new String[]{Manifest.permission.POST_NOTIFICATIONS},
                        PERMISSION_REQUEST_CODE);
            } else {
                startClipboardService();
            }
        } else {
            startClipboardService();
        }

        // Check if started from notification
        if (ClipboardService.ACTION_OPEN_DIALOG.equals(getIntent().getAction())) {
            openClipboardDialog();
        }
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        if (ClipboardService.ACTION_OPEN_DIALOG.equals(intent.getAction())) {
            openClipboardDialog();
        }
    }

    private void startClipboardService() {
        Intent serviceIntent = new Intent(this, ClipboardService.class);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            startForegroundService(serviceIntent);
        } else {
            startService(serviceIntent);
        }
    }

    private void openClipboardDialog() {
        ClipboardDialog dialog = new ClipboardDialog();
        dialog.show(getSupportFragmentManager(), "clipboard_dialog");
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == PERMISSION_REQUEST_CODE) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                startClipboardService();
            } else {
                Toast.makeText(this, "Benachrichtigungsberechtigung erforderlich", Toast.LENGTH_LONG).show();
            }
        }
    }
}

File: app/src/main/java/io/celox/xclip/MainActivity.kt
package io.celox.xclip

import android.Manifest
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.slideInHorizontally
import androidx.compose.animation.slideOutHorizontally
import androidx.compose.animation.togetherWith
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModelProvider
import io.celox.xclip.data.ClipboardEntity
import io.celox.xclip.service.ClipboardService
import io.celox.xclip.ui.ClipboardViewModel
import io.celox.xclip.ui.screen.ClipboardScreen
import io.celox.xclip.ui.screen.DetailScreen
import io.celox.xclip.ui.theme.XClipTheme

class MainActivity : ComponentActivity() {

    private lateinit var viewModel: ClipboardViewModel

    private val permissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestPermission()
    ) { granted ->
        if (granted) {
            startClipboardService()
        } else {
            Toast.makeText(this, getString(R.string.permission_required), Toast.LENGTH_LONG).show()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()

        viewModel = ViewModelProvider(this)[ClipboardViewModel::class.java]

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            permissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
        } else {
            startClipboardService()
        }

        setContent {
            XClipTheme {
                var selectedEntity by rememberSaveable { mutableStateOf<ClipboardEntity?>(null) }

                AnimatedContent(
                    targetState = selectedEntity,
                    transitionSpec = {
                        if (targetState != null) {
                            slideInHorizontally { it } + fadeIn() togetherWith
                                    slideOutHorizontally { -it / 3 } + fadeOut()
                        } else {
                            slideInHorizontally { -it } + fadeIn() togetherWith
                                    slideOutHorizontally { it / 3 } + fadeOut()
                        }
                    },
                    label = "navigation"
                ) { entity ->
                    if (entity != null) {
                        DetailScreen(
                            entity = entity,
                            onBack = { selectedEntity = null },
                            onDelete = {
                                viewModel.deleteEntry(it)
                                selectedEntity = null
                            },
                            onTogglePin = viewModel::togglePin
                        )
                    } else {
                        ClipboardScreen(
                            viewModel = viewModel,
                            onNavigateToDetail = { selectedEntity = it }
                        )
                    }
                }
            }
        }
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        // Notification tap brings app to front - Compose handles it
    }

    private fun startClipboardService() {
        val serviceIntent = Intent(this, ClipboardService::class.java)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            startForegroundService(serviceIntent)
        } else {
            startService(serviceIntent)
        }
    }
}
